### DISCUSSION TITLE
Will pdfcpu work for us?

### DISCUSSION THREAD

**Discussion by gglazer-dustyrobotics** [pdfcpu/pdfcpu#934] [Q&A] [answered]
Hi, gentle folks.

We are looking at using pdfcpu for an architecture-adjacent project and I was hoping you could help me understand a small piece.

Our JSON (with respect to what we want in the PDF) contains very simple primitives like line:

```
                    "line": {
                        "segment": {
                            "end": {
                                "x": -43.33962665938709,
                                "y": -23.230136835586904,
                                "z": 0
                            },
                            "start": {
                                "x": -47.45474018661463,
                                "y": -23.230136835586855,
                                "z": 0
                            }
                        }
                    }
 ```
 
 or arc:
 
 ```
                     "arc": {
                        "endAngle": 3.1138000000000003,
                        "position": {
                            "x": -51.13001121392376,
                            "y": -17.54668728968193,
                            "z": 0
                        },
                        "radius": 0.2286000172090821,
                        "startAngle": 2.370199478664749
                    }
 ```
 
 And what we would like to do is create a JSON file with the suitable PDF headers, iterate over our JSON, adding each element (or some transformation of that element to your schema) to the `pages:1` map. 
 
 Looking at either the commands or the API, it looks like the smallest primitive for create is a box. 
 
 Is this project not pdfcpu suitable or am I missing something?
 
 Thanks for all you do,
 
 Glenn

**Comment by hhrutter** [comment, marked as answer]
Hi there!

Are you saying you are considering drawing positioned arcs and lines via JSON onto the PDF content / user space?
I imagine depending on what you want to draw that the resulting JSON will potentially be huge.

That's also the reason why as of now lines and arcs, circles and what not are not supported by the pdfcpu JSON for creating PDFs.

Check out the current drawing primitives in pdfcpu/draw/draw.go to get a better idea of what's possible.

and of course : pdfcpu/pkg/samples/create/primitives/*.pdf

I just realized that you are talking about drawing in 3D space.
That for sure is not supported right now - we're just rendering onto the 2D PDF user space.

3D annotations might also be smth to consider.
I haven't played around, but it definitely is part of the PDF spec: 13.6.2 PDF annotations

**Reply by gglazer-dustyrobotics** [reply]
Thanks, yes. The files are large: a couple hundred thousand lines or so of JSON. Each one is a line or arc on a blueprint or similar diagram. Thanks for the fast answer, appreciated.
